upgrade undo

edgardoh opened the pull request:
This modifies the history undo to use the same schema than PR  #1800

Now it works on the snapshot of dev->history and a copy of dev->iop, and when is done it replaces it.

It handles the delete of modules on dev->iop, that have to be done when the user undo a new multi-instance or redo a delete multi-instance, as there will be an extra module on dev->iop.
To do this I assume that any multi_priority > 0 must be in history and there must be only one multi_priority = 0 per operation.
If I find a multi_priority > 0 not in history, I delete it.
If I find two multi_priority = 0 for the same operation, one is in history and the other is not, I delete the one that is not in history.

It also handles the move up/down of a multi-instance, and fixes some small issues.

edgardoh commented:
Travis CI is running for a day now, I don't what's going on...

TurboGit commented:
I've tested that, it works great on my side. That's a very nice enhancement/fix. Thanks.

edgardoh commented:
Thanks.

A bit late, but here is my testing:
Add a bunch of multi-instances with parametric mask, reorder them, delete
them all, go crazy with the cntrl+z / cntrl+y
Same with drawn mask and the retouch
Add some style and go back and forward in history.

No crashes and the values are restored ok.

I can't say that I have tested every single scenario, and to keep track of
the values while modules are being reordered, deleted and added is not
easy, but as far as I can tell it works ok.

El jue., 8 nov. 2018 a las 14:33, Pascal Obry (<notifications@github.com>)
escribió:

> I've tested that, it works great on my side. That's a very nice
> enhancement/fix. Thanks.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/1807#issuecomment-437088355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APq1pY19zNCs3fNkRZBNDDDRmaPwCcwuks5utGrwgaJpZM4YSi18>
> .
>
